### DISCUSSION TITLE
Discussion: Testing Features Support

### DISCUSSION THREAD

**Discussion by rayliverified** [ookami-kb/storybook_flutter#42] [Ideas]
Thank you for such a great library which I use heavily in all my projects.

Recently, I've setup golden_toolkit to automatically generate screenshots. However, this doesn't make sense if one is already using this library. The issue is that `golden_toolkit` requires extensive setup and is limited by widget test limitations. Since our UI is already storyboarded, it makes more sense to run tests from this library instead.

Do you have any plans to add testing features? 
There are a few testing features I'm immediately interested in adding to storyboard:
1. Automatically iterate through all stories and capture screenshots. 
2. Automatically generate different screensize previews. 
3. A knob options wrapper for automatically testing different scenarios of knob values. If the knob values are added to a map, they can be automatically iterated through to generate all the combinations of UI states.

What are your thoughts?

**Comment by ookami-kb** [comment]
Thank you, I'm glad you like it!

Yes, it was my plan to add something like that from the very beginning. Unfortunately, I didn't have much time recently, but I'm definitely going to work on that in the (hopefully, near) future. I was thinking about similar use cases, so thanks for putting them into the written form – now we have an issue to start with 😄

**Comment by rayliverified** [comment]
Awesome! I love that we're thinking along the same lines. 

Currently, I don't have too much time to contribute to storybook_flutter but when I do, I would love to contribute. 
Until then, we can make some progress by discussing how the API should look? 

For example, the variants feature is almost fully supported by the Knobs. 
Boolean - true / false. 
Options - enum array.
Slider - min and max bounds.

Missing: 
Text - can be automatically supported by setting text empty or a long paragraph. 

One possible API is that these knobs automatically become variants. 
Additional config over the scenarios to generate can be added by including an additional param such as `scenarioOptions` to the knob. ScenarioOptions could be a class that holds the options.

**Comment by ookami-kb** [comment]
> For example, the variants feature is almost fully supported by the Knobs.
Boolean - true / false.
Options - enum array.
Slider - min and max bounds.

Yes, makes sense.

> Text - can be automatically supported by setting text empty or a long paragraph.

Yes, probably even 3 options by default: empty, short text, long text.

> One possible API is that these knobs automatically become variants.

Yes, I had that in mind as well.

> Additional config over the scenarios to generate can be added by including an additional param such as scenarioOptions to the knob. ScenarioOptions could be a class that holds the options.

I would prefer keeping knobs independent from the tests code, so as an option the test scenario can be generated based on annotations.
